### PR TITLE
[MI-3045]: Added checks for confidential issues at the time of making subscriptions

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
https://github.com/mattermost/mattermost-plugin-gitlab/issues/326

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -544,6 +546,16 @@ func (p *Plugin) subscriptionsAddCommand(ctx context.Context, info *gitlab.UserI
 		return err.Error()
 	}
 
+	if hasPermission := p.permissionToProject(ctx, info.UserID, namespace, project); !hasPermission {
+		p.API.LogWarn("don't have the permission to access the project")
```
Comment (ayusht2810) [inline server/command.go:550]:
```suggestion
		p.API.LogWarn("You don't have permission to access the project")
```

File: server/command.go [hunk 2]
```diff
@@ -544,6 +546,16 @@ func (p *Plugin) subscriptionsAddCommand(ctx context.Context, info *gitlab.UserI
...
+		return "Don't have the permission to access the project"
```
Comment (ayusht2810) [inline server/command.go:551]: same, if changing

File: server/command.go [hunk 3]
```diff
@@ -544,6 +546,16 @@ func (p *Plugin) subscriptionsAddCommand(ctx context.Context, info *gitlab.UserI
...
+	}
+
+	if guestUser := p.checkForGuestUser(ctx, info.UserID, namespace, project); guestUser {
+		p.API.LogWarn("guest user cannot create a subscription")
```
Comment (ayusht2810) [inline server/command.go:555]:
```suggestion
		p.API.LogWarn("Guest user cannot create a subscription")
```

File: server/command_test.go [hunk 4]
```diff
@@ -211,33 +239,87 @@ func TestListWebhookCommand(t *testing.T) {
...
+	}
+
+	p.configuration = &config
+	p.initializeAPI()
+
+	token := oauth2.Token{
+		AccessToken: "access_token",
+		Expiry:      time.Now().Add(1 * time.Hour),
+	}
+	info := gitlab.UserInfo{
```
Comment (ayusht2810) [inline server/command_test.go:299]:
```suggestion
	}
	
	info := gitlab.UserInfo{
```

File: server/command_test.go [hunk 5]
```diff
@@ -211,33 +239,87 @@ func TestListWebhookCommand(t *testing.T) {
...
+		UserID:         "user_id",
+		Token:          &token,
+		GitlabUsername: "gitlab_username",
+		GitlabUserID:   0,
```
Comment (ayusht2810) [inline server/command_test.go:303]: i think we can remove this

File: server/command_test.go [hunk 6]
```diff
@@ -211,33 +239,87 @@ func TestListWebhookCommand(t *testing.T) {
...
+	}
+	encryptedToken, err := encrypt([]byte(config.EncryptionKey), info.Token.AccessToken)
```
Comment (ayusht2810) [inline server/command_test.go:305]:
```suggestion
	}
	
	encryptedToken, err := encrypt([]byte(config.EncryptionKey), info.Token.AccessToken)
```

File: server/command_test.go [hunk 7]
```diff
@@ -211,33 +239,87 @@ func TestListWebhookCommand(t *testing.T) {
...
+
+	api.On("KVGet", "user_id_gitlabtoken").Return(jsonInfo, nil)
+	var subVal []byte
+	api.On("KVGet", "subscriptions").Return(subVal, nil)
+	api.On("LogWarn",
+		mock.AnythingOfTypeArgument("string"),
+		mock.AnythingOfTypeArgument("string"),
+		mock.AnythingOfTypeArgument("string"),
+		mock.AnythingOfTypeArgument("string"),
+		mock.AnythingOfTypeArgument("string")).Return(nil)
```
Comment (ayusht2810) [inline server/command_test.go:321]:
```suggestion
		mock.AnythingOfTypeArgument("string"))
```

File: server/webhook.go [hunk 8]
```diff
@@ -216,6 +217,28 @@ func (p *Plugin) permissionToProject(ctx context.Context, userID, namespace, pro
...
+func (p *Plugin) checkForGuestUser(ctx context.Context, userID, namespace, project string) bool {
+	info, apiErr := p.getGitlabUserInfoByMattermostID(userID)
+	if apiErr != nil {
+		return false
+	}
+
+	result, err := p.GitlabClient.GetProject(ctx, info, namespace, project)
+	if result == nil || err != nil {
+		if err != nil {
+			p.API.LogWarn("can't get project in webhook", "err", err.Error(), "project", namespace+"/"+project)
```
Comment (ayusht2810) [inline server/webhook.go:229]:
```suggestion
			p.API.LogWarn("Can't get project in webhook", "Err", err.Error(), "Project", namespace+"/"+project)
```

File: server/command_test.go [hunk 9]
```diff
@@ -211,33 +239,87 @@ func TestListWebhookCommand(t *testing.T) {
...
 	api.On("KVSet", mock.Anything, mock.Anything).Return(nil)
 	api.On("KVSetWithOptions", mock.AnythingOfType("string"), mock.Anything, mock.AnythingOfType("model.PluginKVSetOptions")).Return(true, nil)
 	api.On("PublishWebSocketEvent", mock.Anything, mock.Anything, mock.Anything).Return(nil)
 
-	p.SetAPI(api)
+	config := configuration{
+		GitlabURL:               "https://example.com",
+		GitlabOAuthClientID:     "client_id",
+		GitlabOAuthClientSecret: "secret",
+		EncryptionKey:           "aaaaaaaaaaaaaaaa",
```
Comment (ayusht2810) [inline server/command_test.go:289]: write some meaningful data

**Review by ayusht2810 — APPROVED** [review]

**Review by manojmalik20 — APPROVED** [review]

File: server/command_test.go [hunk 10]
```diff
@@ -211,33 +239,88 @@ func TestListWebhookCommand(t *testing.T) {
...
+	}
+
 	mockedClient := mocks.NewMockGitlab(mockCtrl)
 	if mockGitlab {
 		mockedClient.EXPECT().ResolveNamespaceAndProject(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return("group", "project", nil)
-		mockedClient.EXPECT().GetProjectHooks(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(hooks, projectHookErr)
-		if projectHookErr == nil {
-			mockedClient.EXPECT().GetGroupHooks(gomock.Any(), gomock.Any(), gomock.Any()).Return(hooks, projectHookErr)
+		if getProjectErr != nil {
+			mockedClient.EXPECT().GetProject(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, getProjectErr)
```
[the remark this answers is not in the document]
Comment (Kshitij-Katiyar) [reply server/command_test.go:254]: copied from the above code.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**